### PR TITLE
[tlcli] Introduce a platform-independent kernel builder

### PR DESCRIPTION
**Pull request by Elarnon**
Currently it is always painful to write generic CLI interfaces to
kernels because of the over-reliance on templated code.  This patch
tries to improve the situation by introducing platform- and
kernel-abstracting APIs in the tlcli library to hide the generic and
allow a value-level API for building the different kernels X platform
combinations.

This is done through the introduction of `Platform`,
`PlatformContextBuilder` and `PlatformContext` abstractions which
perform dispatching over the platform (currently x86 or cuda), and a
`KernelBundle` abstraction which packs the relevant information about a
built kernel, namely the candidates, a checking function (for output
validity), and a reference function (for performance comparisons).

### CODE REVIEW

**Comment by andidr** [comment]
> Currently it is always painful to write generic CLI interfaces to
> kernels because of the over-reliance on templated code. This patch
> tries to improve the situation by introducing platform- and
> kernel-abstracting APIs in the tlcli library to hide the generic and
> allow a value-level API for building the different kernels X platform
> combinations.

Not sure what is meant by "value-level" API.

> This is done through the introduction of `Platform`,
> `PlatformContextBuilder` and `PlatformContext` abstractions which
> perform dispatching over the platform (currently x86 or cuda), and a
> `KernelBundle` abstraction which packs the relevant information about a
> built kernel, namely the candidates, a checking function (for output
> validity), and a reference function (for performance comparisons).

**Comment by Elarnon** [comment]
Value level API is in opposition to the previous type level API; i.e. we have values to represent both platforms and kernels.  Before we would need to use type parameters for both everywhere, leading to long compile times and somewhat complex templates.  This PR encapsulates that complexity in the `KernelBundle` and `Platform` family of types.

File: telamon-cli/src/bin/cuda_search/main.rs [hunk 1]
```diff
@@ -68,118 +17,66 @@ struct Opt {
...
+                bundle.candidates,
+                Some(&bundle.check_fn),
+            )
+            .unwrap_or_else(|| panic!("no candidates found for kernel {}", kernel));
+
+            let best_fn = codegen::Function::build(&best.space);
+            let runtime = context.benchmark(&best_fn, args.num_code_runs);
+
+            let ref_runtime = Bench::default()
+                .warmup(4)
```
Comment (andidr) [inline telamon-cli/src/bin/cuda_search/main.rs]: Make this a parameter or use default value

File: telamon-cli/src/lib.rs [hunk 2]
```diff
@@ -357,6 +354,90 @@ mod cuda_reference {
...
+    pub struct X86Reference {
+        _priv: (),
+    }
+
+    impl<'a> Reference<'a, linalg::Axpy<'a, f32>> for X86Reference {
+        type Context = telamon_x86::Context;
+
+        fn eval_reference(&self, _params: &(i32, bool), _context: &Self::Context) -> f64 {
+            // TODO
+            1.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: Rather `panic!` instead of returning an arbitrary value

File: telamon-cli/src/lib.rs [hunk 3]
```diff
@@ -357,6 +354,90 @@ mod cuda_reference {
...
+    impl<'a> Reference<'a, linalg::MatVec<'a, f32>> for X86Reference {
+        type Context = telamon_x86::Context;
+
+        fn eval_reference(
+            &self,
+            _params: &(i32, i32, bool),
+            _context: &Self::Context,
+        ) -> f64 {
+            // TODO
+            1.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: Same here

File: telamon-cli/src/lib.rs [hunk 4]
```diff
@@ -357,6 +354,90 @@ mod cuda_reference {
...
+    impl<'a> Reference<'a, linalg::Gesummv<'a, f32>> for X86Reference {
+        type Context = telamon_x86::Context;
+
+        fn eval_reference(
+            &self,
+            _params: &(i32, i32, bool),
+            _context: &Self::Context,
+        ) -> f64 {
+            // TODO
+            1.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: And here

File: telamon-cli/src/lib.rs [hunk 5]
```diff
@@ -357,6 +354,90 @@ mod cuda_reference {
...
+    impl<'a> Reference<'a, linalg::FusedMM<'a, f32>> for X86Reference {
+        type Context = telamon_x86::Context;
+
+        fn eval_reference(
+            &self,
+            _params: &linalg::FusedMMP,
+            _context: &Self::Context,
+        ) -> f64 {
+            // TODO
+            1.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: and here

File: telamon-cli/src/lib.rs [hunk 6]
```diff
@@ -357,6 +354,90 @@ mod cuda_reference {
...
+    impl<'a> Reference<'a, linalg::BatchMM<'a, f32>> for X86Reference {
+        type Context = telamon_x86::Context;
+
+        fn eval_reference(
+            &self,
+            _params: &linalg::BatchMMP,
+            _context: &Self::Context,
+        ) -> f64 {
+            // TODO
+            1.
```
Comment (andidr) [inline telamon-cli/src/lib.rs]: and here

**Pull request merged**